pint/quantity.py

```diff
@@ -1486,12 +1498,14 @@ def __eq__(self, other):
                     else:
                         raise OffsetUnitCalculusError(self._units)
 
-            return self.dimensionless and eq(
-                self._convert_magnitude(self.UnitsContainer()), other, False
-            )
+            if self.dimensionless:
+                return eq(self._convert_magnitude(self.UnitsContainer()), other, False)
 
+            return bool_result(False)
+
+        # TODO: this might be expensive. Do we even need it?
         if eq(self._magnitude, 0, True) and eq(other._magnitude, 0, True):
-            return self.dimensionality == other.dimensionality
+            return bool_result(self.dimensionality == other.dimensionality)
```

Comment:
I can't figure out what this block is for. I tried removing the whole block and rerun the tests to find out why this is necessary, but this didn't break any tests. Since this is potentially expensive (we might end up comparing the values multiple times on huge arrays) I'd suggest removing this:
```suggestion
```